fscbco/topnews
# Fetch, Display, and Allow Users to Like Stories From Hacker News

## What Was Done
- Add User sign in and sign out to the app. I probably didn't need to pull in the Devise views but it doesn't hurt since we can update these at a future time.
- Added the `Story` Model.
  - I opted to make this a table for performance reasons and because I'd inevitably have to create some kind of record in order to relate likes back to the story responses from the API. This seemed like the cleanest way to do this.
- Added the `FeedController` and model.
  - I had trouble getting `store_accessor` to work so I opted to create methods for the values I wanted to show on the front end. 
  - The `FeedController` utilizes the `FetchTopStoriesFromApiService` which pulls in the ids for the top stories and caches them for 15 minutes and also uses `FindOrCreateStoryService` which grabs and saves data for stories
  - Added the `LikedStories` Controller and views. This pulls in all of the stories that were liked in no particular order. It shares a partial with the `FeedController`. 
- I'm sharing the `FeedHelper` between the `LikedStoriesController` and `FeedController` since they are both using the same partial. This partial should probably be extracted out into a more agnostic folder. 
- I made all of the net new views(Likes, LikedStories, and Feed) only accessible to users who are logged in mostly because of the like logic. I could always just hide the like button if the user isn't logged in. 
- Added a `LikesController` and model. This is a simple join table for now. 
- Added VCR for recording tests.
- Added HTTParty to fetch APIs easier.
- Added Guard for quality of life while writing tests.
## Screenshots
### /feed
![image](https://github.com/fscbco/topnews/assets/13066920/6ff79ed7-77b5-41ca-a81e-3cdea373d6bb)

### /liked_stories
![image](https://github.com/fscbco/topnews/assets/13066920/ab320c40-a660-4480-a1d8-875e9e7cb83f)


## To Test
- Verify you can't access `/feed` and `/liked_stories` without being logged in. 
- Create an account
- In the console, update your first and last name(I wasn't able to get to this part)
- Navigate to `/feed`, like some stories. Verify they show up in the `/liked_stories` path. 
- Log into or create another user. Like some stories, both by unliked ones and ones that the other user has liked. 
- Verify they show up in `/liked_stories`
- Verify both views show the users who liked the stories.
